Replace strata with time-weighted, leave-one-out, rolling building average

dfsnow opened the pull request:
This PR completely replaces our building strata features with a more performant, more predictive, more empirically sound rolling mean feature. Specifically, we construct the 5-year rolling average for each condo building, excluding the current sale and weighting by time, then use it as a predictor in the primary unit-level model.

This improves performance, has simpler code, and aligns more closely with how most people assume their condo unit is valued.

Closes #7, #21, #73, #93, #94.

dfsnow reviewed (commented):
Ditching KNN imputation for bag imputing here since it's faster and can use categorical inputs. Note that the `trees` var becomes part of CV now, so we'll have to re-tune this model.

dfsnow commented on R/recipes.R:
Ditching KNN imputation for bag imputing here since it's faster and can use categorical inputs. Note that the `trees` var becomes part of CV now, so we'll have to re-tune this model.

dfsnow commented on README.Rmd:
Changes in here are just updating/removing references to strata + fixing typos.

dfsnow commented on misc/desk_review_template.xlsx:
Not visible in the diff, but I dropped the strata and flag columns and replaced them with the building mean and count of sales used to construct the mean.

dfsnow commented on pipeline/00-ingest.R:
```diff
@@ -164,35 +109,19 @@ training_data <- dbGetQuery(
-  WHERE condo.year
-      BETWEEN '{params$input$min_sale_year}'
-      AND '{params$input$max_sale_year}'
+  WHERE CAST(condo.year AS int)
+      BETWEEN CAST({params$input$min_sale_year} AS int) -
+        {params$input$n_years_prior}
+      AND CAST({params$input$max_sale_year} AS int)
```
Like in res, we were accidentally truncating the sales sample we needed to construct lagged features, so this lets in all sales after 2011, then truncates the training data after the rolling building means are constructed later in the script.

dfsnow commented on pipeline/00-ingest.R:
```diff
@@ -164,35 +109,19 @@ training_data <- dbGetQuery(
 )
 tictoc::toc()
 
-# Raw sales document number data used to identify some sales accidentally
-# excluded from the original training runs. See
-# https://github.com/ccao-data/data-architecture/pull/334 for more info
-tictoc::tic("Sales data pulled")
-sales_data <- dbGetQuery(
-  conn = AWS_ATHENA_CONN_NOCTUA, glue("
-  SELECT DISTINCT
-      substr(saledt, 1, 4) AS year,
-      instruno AS doc_no_old,
-      NULLIF(REPLACE(instruno, 'D', ''), '') AS doc_no_new
-  FROM iasworld.sales
-  WHERE substr(saledt, 1, 4) >= '{params$input$min_sale_year}'
-  ")
-)
-tictoc::toc()
-
```
This was a remnant of the fix from the tail end of last season and is no longer needed.

dfsnow commented on pipeline/00-ingest.R:
This is the actual building mean construction. The code here isn't intuitive _at all_ but its the only way to get it done fast. It helps a lot to just look at the `bldg_rolling_means_dt` dataframe in the viewer so you can follow each step of the construction.

dfsnow commented on R/recipes.R:
Ditching KNN imputation for bag imputing here since it's faster and can use categorical inputs. Note that the trees var becomes part of CV now, so we'll have to re-tune this model.

dfsnow commented on README.Rmd:
Changes in here are just updating/removing references to strata + fixing typos.

dfsnow commented on misc/desk_review_template.xlsx:
Not visible in the diff, but I dropped the strata and flag columns and replaced them with the building mean and count of sales used to construct the mean.

dfsnow commented on params.yaml:
```diff
@@ -224,8 +212,7 @@ model:
       - "shp_parcel_num_vertices"
-      - "meta_strata_1"
-      - "meta_strata_2"
+      - "meta_pin10_bldg_roll_mean"
```
I didn't do it here, but think it may also be worth including the _count_ of sales as a feature, with the thinking being that high turnover = lots of flips = higher prices.

jeancochrane replied:
Note that this is now done on the line below.

wrridgeway reviewed (approved):
Other than the insane data.table bit it looks great to me. Inspired stuff.

dfsnow commented on params.yaml:
```diff
@@ -238,21 +225,24 @@ model:
       - "char_yrblt"
+      - "meta_township_code"
+      - "char_land_sf"
+      - "char_building_units"
```
Fixed!

dfsnow commented on README.md:
```diff
@@ -360,9 +322,9 @@ have recent sales.
-look at nearby buildings that have similar unit counts/age and then try
-to assign an appropriate strata to the target building.
+assessment, the building’s mean feature is imputed. The model will look
+at nearby buildings that have similar unit counts/age and then try to
```
Updated in ffcfdec

jeancochrane commented on pipeline/00-ingest.R:
**[Question, non-blocking]** Looking at the `bldg_rolling_means_dt` output, the `cnt` value doesn't always seem to match my expectation. I expect `cnt` to reflect the number of non-outlier sales in the building within the rolling window prior to the current "sale". It seems like in some cases, the count is off by one. Take PIN10 `0101101030` for example:

```R
bldg_rolling_means_dt %>%
    filter(meta_pin10 == '0101101030', data_source == 'training') %>%
    select(meta_pin10, meta_pin, meta_sale_date, cnt)
```

Here's the output:

```
Key: <meta_pin10, meta_sale_date>
   meta_pin10       meta_pin meta_sale_date   cnt
       <char>         <char>         <Date> <num>
1: 0101101030 01011010301080     2015-12-01     0
2: 0101101030 01011010301078     2018-01-01     1
3: 0101101030 01011010301082     2018-09-01     2
4: 0101101030 01011010301069     2019-08-01     3
5: 0101101030 01011010301085     2020-03-01     4
6: 0101101030 01011010301075     2022-05-04     5
7: 0101101030 01011010301067     2022-06-15     6
8: 0101101030 01011010301069     2024-04-01     5
```

In this case, it seems like `cnt` is counting some sales that are outside the 5-year rolling window; e.g. if we go 6 positions back from sale 7, we get sale 1, which took place more than 5 years before sale 7. Does that make sense? Or am I misinterpreting the way `cnt` is supposed to work?

dfsnow replied:
Oh nice catch. This was caused by the window size being relative to the current date rather than the lagged date. I fixed it in 02d0df1. It's probably worth another quick review just to be sure I've got the logic right. I added the window size as a variable in `bldg_rolling_means_dt` for easier debugging.

jeancochrane reviewed (approved):
Really nice work! I played around with `bldg_rolling_means_dt` a bunch and it mostly seems to match my expectation, aside from one question about `cnt` that I raised below. Looking forward to seeing the stats on this approach once CV is done.

dfsnow commented on params.yaml:
```diff
@@ -60,26 +60,14 @@ input:
-    # Number of quantiles of centers to use when constructing strata
-    k_1: 10
-    k_2: 100
+  # Rolling time window size for building mean feature
+  n_years_prior: 5
```
I'd say it's a parameter of the input data in general, not just the building feature, since it's also used in the initial sales fetch.

dfsnow commented on README.Rmd:
```diff
@@ -272,24 +241,24 @@ The CCAO faces a number of ongoing issues specific to condominium modeling. We a
 ### Buildings Without Sales
 
-When no sales have occurred in a building in the 5 years prior to assessment, the building's strata features are imputed. The model will look at nearby buildings that have similar unit counts/age and then try to assign an appropriate strata to the target building.
+When no sales have occurred in a building in the 5 years prior to assessment, the building's mean feature is imputed. The model will look at nearby buildings that have similar unit counts/age and then try to assign an appropriate average to the target building.
```
Fixed in 9eefd77!

dfsnow commented on README.Rmd:
```diff
@@ -303,11 +272,7 @@ As with the [residential model](https://github.com/ccao-data/model-res-avm), the
-**Q: How do I see my condo building's strata?**
-
-Individual building [strata](#condo-strata) are not included with assessment notices or shown on the CCAO's website. However, strata *are* stored in the sample data included in this repository. You can load the data ([`input/condo_strata_data.parquet`](./input/condo_strata_data.parquet)) using R and the `read_parquet()` function from the `arrow` library.
+* Other sales in the building. This is captured by a rolling average of sales in the building over the past 5 years.
```
Fixed in 9eefd77!

wagnerlmichael commented on params.yaml:
```diff
@@ -313,7 +303,7 @@ model:
       lambda_l1: 0.697
       lambda_l2: 0.002
-      neighbors: 15
+      trees: 25
```
**[Question, non-blocking]:** How did you end up at 25 here?

dfsnow replied:
It's just the default for that recipe step. We'll change it once we do a full CV run and find the optimal number.

dfsnow commented on pipeline/00-ingest.R:
```diff
@@ -457,53 +457,60 @@ bldg_rolling_means_dt[
   # from the past N years
+  !sv_is_outlier & meta_modeling_group == "CONDO" | data_source == "assessment",
+  window_size := seq_len(.N) -
+    findInterval(meta_sale_date %m-% offset, lag_sale_date[-1]) - 1,
```
Here's the fix for the count issue @jeancochrane. This makes the index interval compare the current date to the lagged date. So looking at something like this:

| row | sale_date  | lagged_date | window_size | count |
|-----|------------|-------------|-------------|-------|
| 1   | 2015-01-01 | NA          | 0           | 0     |
| 2   | 2023-01-01 | 2015-01-01  | 0           | 0     |
| 3   | 2024-01-01 | 2023-01-01  | 1           | 1     |

`findInterval` looks for `lagged_date` values that are >= `sale_date` - 5 years, excluding the first row (which can never have a lagged date). A good example building to look at is `2716210034`.

dfsnow commented:
@wrridgeway @jeancochrane This is now ready for re-review! I greatly simplified the mean calculation logic, added better commenting, and fixed some counting issues. I recommend taking a look at just the ingest script and diffing against `2025-assessment-year`, not your last review.

jeancochrane commented on pipeline/00-ingest.R:
**[Question, non-blocking]** What are the different cases in which we might expect a mean less than zero, a null mean, or an infinite mean? Some that come to mind:

1. Sales in the window have amounts that are incorrectly < 0 ➡️ Negative mean
2. No sales in the window ➡️ Infinite mean

My thought is that we might want to fail loudly for cases that point to problematic data (case 1), whereas it could be fine to coerce to null in cases where we expect weirdness (case 2). This would point to splitting these cases out into separate tests.

No strong feelings here! Just thinking out loud.

dfsnow replied:
I didn't actually see any < 0 mean cases, was just trying to be defensive here. But I think you're right it's worth throwing a warning or error if any sales or means are < 0. I'll add a check just below here for any such cases.

There are a bunch of things that make inf/NaN values here, including no (arms-length, market) sales in the window, being the very first sale in a building, being the _only_ sale in the building, etc. I agree we're safe to coerce these cases to NA.

dfsnow replied:
Just kidding, there's actually some floating point nastiness that causes a few negative means:

![Screenshot 2025-02-07 at 12 20 14 PM](https://github.com/user-attachments/assets/4da7224e-773e-48ba-ab34-81a58d0a79e6)

Where the first to columns are the numerator and denominator of the mean, respectively.

However, it seems like the issues is totally resolved by just setting the `algo` argument in the `froll*` functions to `"exact"`, which I did in e4a82b8.

jeancochrane commented on pipeline/00-ingest.R:
**[Suggestion, non-blocking]** Thank you for the clear example of the interval math! I was tripped up a bit thinking there was an off-by-one error due to my own forgetting that we exclude the target sale price later on in the transformation, so it might be helpful to hint at that here by adding one more paragraph at the end:

```R
  #
  # In the case of the 4th element of Y, we end up with a window size of 4 - 2 == 2.
  # This means that our rolling mean will include the last two sales in Y, the sales
  # at positions 3 and 4. Since position 4 is the target sale, we will avoid data
  # leakage later on in the pipeline by subtracting the target sale price from the
  # mean (or subtracting 0 in the case of the assessment set, which does not
  # have a sale price).
```

dfsnow replied:
Good thinking. Added in 50f2358!

jeancochrane reviewed (approved):
Very nice work 👏🏻

wrridgeway commented:
Thanks for updating the commenting. Super clear what's going on now.
